### ISSUE TITLE
Cannot restart run with model

### ISSUE THREAD

**Issue by Higginbottom** [sirocco-rt/sirocco#508]
Trying to get py/pluto running with a supplied SED is causing problems.
I use the following model files

[model.ls.txt](https://github.com/agnwinds/python/files/3134030/model.ls.txt)
[4.0pc_soft_py.dat.txt](https://github.com/agnwinds/python/files/3134031/4.0pc_soft_py.dat.txt)

The initial run, using this input file

[input.pf.txt](https://github.com/agnwinds/python/files/3134042/input.pf.txt)

with this hydro model file

[00000001.pluto.txt](https://github.com/agnwinds/python/files/3134033/00000001.pluto.txt)

runs fine with the command py83a -z input

However, if you then try and restart it by changing the number of ionization cycles to 5, and doing

runs fine with the command py83a -z -r  input

it doesn't get far, stopping with errors:

emittance_contiuum: freqmin 4.840000e+17 freqmax 2.420000e+18
emittance_contiuum: emin 2.001921e+03 emax 1.000960e+04
emittance_continum: Requested wavelengths extend beyond models wavelengths for list 
lambda 1.238812 6.194060  model 0.000000 0.000000
agn_init: Supplied model SED has no power in 2-10keV band - cannot continue

which is odd....

investigating...

**Comment by Higginbottom** [comment]
get_models isn't run on restart 

so, is the bug that it isn't run, or that the models aren't included in the wind save file.....

**Comment by Higginbottom** [comment]
Same problem occurs with cv_kur.pf in examples. Well similar - clearly looks like it doesn't have models....

The question is - did it ever work???

**Comment by Higginbottom** [comment]
Simply adding 


  n += fwrite (&mods, sizeof (mods), 1, fptr);
  n += fwrite (&comp, sizeof (comp), 1, fptr);

and


  n += fread (&mods, sizeof (mods), 1, fptr);
  n += fread (&comp, sizeof (comp), 1, fptr);

to the relevant parts of windsave.c allow it to run - however the windsave file balloons,  need to make it more intelligent so it outputs only the actual filled parts of the arrays....

Also need to res to make sure it is really working..

**Comment by Higginbottom** [comment]
Seems OK - but the total size of the models structure is 500Mb - so I really need to work out how to shrink it.... To be frank - its rather large to be held in RAM too!!!

Should look into making the models structure dynamic...

**Comment by Higginbottom** [comment]
Looking into the possibility of reading in the models from scratch after a restart - so the wind save file stays small and there is also the chance of reading in a new model - it would be messy, since one would have to check all the components for a model type spectrum. Then read them in possibly different models for different components. The wind save option seems best...

**Comment by kslong** [comment]
Nick, I am trying to understand this thread.  Is it simply that the models are not read in on restarts, or is there anything more to it than that?  

If it is just that the models are not read in, then I agree with your solution that we should just make that happen.  The models should NOT be part of the windsave file.

**Comment by Higginbottom** [comment]
I think it is just that - models aren't read in.

I'm pretty agnostic about whether to put the models into the wind save or read them in. Neither is trivial and because of some changes made a while ago the memory needed to hold the models is pretty huge - and hard wired. So I think making this dynamic is a good idea anyway. But again this isn't easy because in general the size of the required arrays is only known once they are all read in.

Reading models in on restart will require a working out which components are using models - which will necessitate messy code to check disks, stars and agn separately ....

**Comment by kslong** [comment]
I am NOT agnostic on this question.  We do not store the atomic data, we read it in.  We should not store the models either.  As you have already noted the windsavefiles become quite large, and 99% of the time we do not need to.

I do recognize that it is a bit tricky, but suspect this is pretty much the case in any event.  We may need to keep a skeleton of what was read in in a structure, so the models are counted, but I suspect we just need some code that reads a bit more of the .pf file than we may do at present.

**Comment by Higginbottom** [comment]
Fair enough - I note that there is a thing called 'model_list' stored in geo - this might be the way into the problem. As long as the correct pointers into the models array are stored...

**Comment by kslong** [comment]
There are at least two ways to go here:


  *   Reread portions of the .pf file and use this to reconstruct what needs to back in.
  *   Save in the windsavefile the ModSum structures, that is comps.  Maybe adding one or two variables and use this the read everything back without using the .pf file

Note that restarts are related a bit to system type previous.  These do read the .pf file in, or should.  We should check (a) that they do read the models in correctly, and (b) decide whether this means the the .pf file approach is favored.

**Comment by kslong** [comment]
I suspect that this worked properly in the past, but that it may have been broken when I rearranged inputs last year to create a more logically ordered input file.

**Comment by kslong** [comment]
OK, NIck, I have looked into this a bit, and I think this may be fairly straightforward.   Maybe you have already analyzed is going on as well.

The models are read in as part of get_spectype which is a generic function which is called multiple times once basically for each type of radiation source (star, disk, etc.)   On restarts, all of the section of the inputs is omitted.

This is the portion of the code in get_spect_type where the models are read in 

```
    if (*spectype == SPECTYPE_MODEL)
    {
      if (geo.run_type == RUN_TYPE_PREVIOUS)
      {                         // Continuing an old model
        strcpy (model_list, geo.model_list[get_spectype_count]);
      }
      else
      {                         // Starting a new model
        strcpy (model_list, get_spectype_oldname);
      }
      rdstr ("Input_spectra.model_file", model_list);
      get_models (model_list, 2, spectype);

      strcpy (geo.model_list[get_spectype_count], model_list);  // Copy it to geo
      strcpy (get_spectype_oldname, model_list);        // Also copy it back to the old name
      get_spectype_count++;
    }

```
(Aside, there is something odd about the code above.  I don't for system type PREVIOUS it should be necessary to have the same spectypes ... that's really what one wants for RESTARTS, but this code is never called for restarts). 

The information about what model list one needs to read in for each radiation source is contained , in the geo structure in the array 
```

  /* The next set pf variables assign a SPECTYPE (see above) for
     each possible source of radiation in a model.  The value assigned can be different for
     the ionization and detailed spectrum generation part of the code */

  int star_ion_spectype, star_spectype; /* The type of spectrum used to create the continuum
                                           for the star in the ionization and final spectrum calculation */
  int disk_ion_spectype, disk_spectype; /* Same as above but for the disk */
  int bl_ion_spectype, bl_spectype;     /* Same as above but for the boundary layer */
  int agn_ion_spectype, agn_spectype;   /* Same as above but for the AGN */
  int search_light_ion_spectype, search_light_spectype; /* Same as above but for the search_light. Created for 1d test */

  char model_list[NCOMPS][LINELENGTH];  /* The file which contains the model names and the associated values for the model */

```

Every time models are too, be read in an element of the geo..model_list is read in, but this means that there can be several lines that say the same model list because we could have multiple components that use the same models.   Indeed, if you only have one component, but use Kurucz models for both ionization and detailed spectra it will appear here.  

Now the spectypes above will be negative for things like BB or brems, and 0 or positive for models.  

So this means in principle that if we want to read the model data in,  we just need to call get_models for all of the model lists, in the order that they have been stored in the model_list array.

The only issue is that as far as I can tell we do not keep the total number of actual model lists we need to read in.

One could figure it out from looking for the maximum value in all of the various spectypes, but an easier method might be either (a) to add a number to geo that tells you the number of model lists that have to be read in, or (b) to initialize all of the vlues in model list to something, e.g "None", and read in model lists until your get to this keyword or the maximum number that are allowed.  

I can try this in the next day or so, or you can do it, I am sure.  What's your pleasure?

**Comment by Higginbottom** [comment]
It's been a looooong time since this worked if ever. I've gone back quite a long way - back to the point in 2016 where libfitsio was removed from the repo and I can no longer get python to compile - ad the same problem occurs. So I cant get help from the past.

I've been looking at the same code - I agree that using model_list is sensible. I was looking at the code yesterday - one small issue is that the array gets new models put into it - even if the code later decides the model is a copy of an earlier one - so I just need to ensure that the army only retains real models. Then one could either populate the array initially with something that would tell the code it has reached the end of 'real' data - or as you say have an extra 'nmods' parameter that says how long the array is.

I do think we should make these model arrays dynamic - the sizes were massively increased a while back - meaning that each model will take up 500Mb of RAM - this could easily cause problems... My models are *tiny* so rattle around in such huge arrays...

**Comment by Higginbottom** [comment]
OK, I think I have a fix - just testing.
Essentially what I have done is 
1: Made a global variable called geo.model_count which says how long the array geo.model_list is.
2: I have essentially shifted the test for duplicate models up into setup.c - so it only appends a new model to geo.model_list if it is new. Previously the model wasn't read in, due to a test in get_models, but it was appended to the array anyway.
3: I have a small extra test in python.c which basically asks if geo.model_count >0, if so it calls get_model with each of the models listed in geo.model_list. The correct model numbers relating to each component should already be read in - so it is just a case of reading in the models. I'm now testing...

I will also look into making the model arrays dynamic, I'm sure that I'm going to run into problems otherwise, and it is crazy to have huge static arrays held in memory...

**Comment by kslong** [comment]
Sounds good.  I do not believe the current version of the code reads the data in multiple times.  In fact I am quite sure it does not.

**Comment by kslong** [comment]
I must have missed your comment 2 above incorrectly.  I guess I still don't understand what you meant.

Dynamically allocating the arrays seems to me necessary only if they fill up a lot of space, but certainly it is  fine to do this.

**Comment by Higginbottom** [comment]
initial problem fixed with pull #509 
Will set up a new bug for resizing model arrays..
